374 add route warning missing

NejcS opened the pull request:
This PR fixes the problem that the user does not have an option to add a warning to a route. The reason is that the warnings are listed at the top of the route page where they do not have a button. This PR is a proposal of how we can fix this (we talked about this solution briefly on a video call).

It merges all the "Dodaj komentar", "Dodaj opis smeri" and "Dodaj opis razmer" buttons into one "Dodaj komentar". After clicking this, the user can select the type of comment that they want to submit in the comment form. The "Opozorilo" option enables the user to submit route warnings.

This solution is a proposal so we might have to debate about it.

Before:
<img width="1085" alt="image" src="https://user-images.githubusercontent.com/6022408/181763418-3e8f9f6b-02be-4e49-aa6e-c24da2052484.png">

After:
<img width="1120" alt="image" src="https://user-images.githubusercontent.com/6022408/181763479-61a787b5-5671-45ef-91e8-8b19b4ab409f.png">

Before:
<img width="1076" alt="image" src="https://user-images.githubusercontent.com/6022408/181763613-5d526153-bca0-42d4-a9e9-5bf805dcd862.png">

After:
<img width="1091" alt="image" src="https://user-images.githubusercontent.com/6022408/181765342-f5c0803d-dd63-4f4a-bce8-e4ec23dc82b6.png">


Example of comment form with comment type dropdown:
<img width="653" alt="image" src="https://user-images.githubusercontent.com/6022408/181764623-44401282-96f1-493a-9b00-9a6a9a5dfd64.png">

NejcS commented on src/app/shared/components/comment-form/comment-form.component.ts:
```diff
@@ -34,9 +33,11 @@ export class CommentFormComponent implements OnInit {
     content: new FormControl(),
   });
 
+  minDate = new Date();
```
I think we should prevent the user from submitting a warning with an `exposedUntilDate` set to a past date.

salamca replied:
👍

NejcS commented on src/app/shared/components/comment-form/comment-form.component.ts:
```diff
@@ -48,44 +49,61 @@ export class CommentFormComponent implements OnInit {
-    this.commentForm.addControl('exposedUntil', new FormControl());
+    this.commentForm.addControl(
+      'exposedUntil',
+      new FormControl(null, [Validators.required])
```
Prevent the user from submitting the form without the `exposedUntil` date. The endpoint fails anyway and the displayed message in this case is not helpful.

salamca replied:
We should let the user post warnings without exposing them on home page. -> so this field should stay not required...
maybe additional explanation (in a hint) would be welcome. or even a checkbox?

why BE fails? probably this is a FE bug that I suspect was introduced when migrating from momentjs to dayjs. dayjs does not return null when null is fed to it...

salamca replied:
resolved this bug with: https://github.com/plezanje-net/web/pull/402/commits/8e575ae87fdfb77659d1fc737c0a720acaaee670

NejcS commented on src/app/shared/components/comment-form/comment-form.component.ts:
```diff
@@ -48,44 +49,61 @@ export class CommentFormComponent implements OnInit {
 
+  removeExposedUntilField() {
+    this.commentForm.removeControl('exposedUntil');
+  }
```
This form control now also needs to be removed because it is a required field of this form. If not removed, it would prevent the user from submitting in case they first selected a `warning` type and then some other type.

salamca replied:
Also resolved with https://github.com/plezanje-net/web/pull/402/commits/8e575ae87fdfb77659d1fc737c0a720acaaee670

But we should anyway hide the control (and empty it) in case that we are not posting a warning.

I wonder if hiding it instead of removing it would make for less of flickering when changing comment type in the popup.

NejcS commented on src/app/shared/editor/editor.component.ts:
```diff
@@ -36,6 +37,7 @@ export class EditorComponent implements OnInit {
     toolbar:
       'undo redo | bold italic underline strikethrough | \
       bullist numlist | link',
+    elementpath: false,
```
This removes the circled element:
<img width="658" alt="Screenshot 2022-07-29 at 15 02 47" src="https://user-images.githubusercontent.com/6022408/181766233-b2de9f68-e090-4aca-b37c-bf4e63a317ec.png">

NejcS commented on src/app/shared/components/comment-form/comment-form.component.ts:
```diff
@@ -48,44 +49,61 @@ export class CommentFormComponent implements OnInit {
 
   ngOnInit(): void {
+    this.updateFormType();
     if (this.data.comment != null) {
```
Editing a comment is currently not possible anyway but I didn't want to get rid of this code. It will be useful once we start working on https://github.com/plezanje-net/web/issues/375

NejcS commented on src/app/shared/components/warnings/warnings.component.scss:
```diff
@@ -4,7 +4,14 @@
+    p:last-of-type {
+      margin-bottom: 0;
+    }
+  }
```
Before:
<img width="1056" alt="image" src="https://user-images.githubusercontent.com/6022408/183098878-68957379-2969-4308-a2fe-21f2ffd55f01.png">


After:
<img width="1070" alt="image" src="https://user-images.githubusercontent.com/6022408/183098762-40e45b6d-15c9-413b-9e1c-bbc8a70767d0.png">

salamca replied:
👍

bonus: could we avoid ::ng-deep somehow?

NejcS replied:
No, because the paragraph is the content of the comment and not part of the warnings component.

NejcS commented:
After debating how to improve this mess, we decided to remove the option to add new conditions comments and descriptions. This way, the dropdown for selecting the type of comment the user wants to submit is not needed anymore and removed. To add warnings, I added a "menu" button to the header of the route page. It is now the same as it is on the crag page where you can also add warnings from the same menu.

<img width="1289" alt="Screenshot 2022-08-05 at 16 19 39" src="https://user-images.githubusercontent.com/6022408/183099115-67021daa-2538-4ab9-9af4-3f5ebe6ae2e8.png">

salamca commented:
Editing or deleting a comment throws an error>
<img width="630" alt="image" src="https://user-images.githubusercontent.com/10534304/185589124-09c5045c-0980-403f-ba86-34bf100b1245.png">

salamca commented:
Content of a comment should be required. 
(posting an empty comment produces an error)
Best to fix this along the way...

NejcS commented:
@salamca the content field of the comment form is now required. I also fixed the query error that was thrown on edit and deleted.
